tik/lib/CartographerKernel.cpp

```diff
@@ -291,500 +281,926 @@ namespace TraceAtlas::tik
+                                                prevSel = sel;
+                                            }
+                                            for (unsigned int j = 0; j < phi->getNumIncomingValues(); j++)
+                                            {
+                                                if (phi->getIncomingBlock(j) == sw->getParent())
+                                                {
+                                                    phi->removeIncomingValue(j);
+                                                }
+                                            }
+                                            phi->addIncoming(prevSel, sw->getParent());
+                                        }
+                                    }
                                 }
                             }
                         }
                     }
                 }
-                if (exportMap.find(i) != exportMap.end())
+            }
+        }
+    }
+
+    void CartographerKernel::InlineFunctionsFromBlocks(std::set<int64_t> &blocks)
+    {
+        bool change = true;
+        while (change)
+        {
+            change = false;
+            for (auto fi = KernelFunction->begin(); fi != KernelFunction->end(); fi++)
+            {
+                auto baseBlock = cast<BasicBlock>(fi);
+                auto id = GetBlockID(baseBlock);
+                if (blocks.find(id) == blocks.end())
+                {
+                    continue;
+                }
+                for (auto bi = fi->begin(); bi != fi->end(); bi++)
                 {
-                    Instruction *buildBase;
-                    if (isa<PHINode>(i))
+                    if (auto ci = dyn_cast<CallBase>(bi))
                     {
-                        buildBase = block->getFirstNonPHI();
+                        if (auto debug = ci->getMetadata("KernelCall"))
+                        {
+                            continue;
+                        }
+                        auto id = GetBlockID(baseBlock);
+                        auto info = InlineFunctionInfo();
+                        auto r = InlineFunction(ci, info);
+                        SetBlockID(baseBlock, id);
+                        blocks.insert(id);
+                        if (r)
+                        {
+                            change = true;
+                        }
+                        break;
                     }
-                    else
+                }
+            }
+        }
+        //erase null blocks here
+        auto blockList = &KernelFunction->getBasicBlockList();
+        vector<Function::iterator> toRemove;
+
+        for (auto fi = KernelFunction->begin(); fi != KernelFunction->end(); fi++)
+        {
+            if (auto b = dyn_cast<BasicBlock>(fi))
+            {
+                //do nothing
+            }
+            else
+            {
+                toRemove.push_back(fi);
```

Comment:
what are these non-blocks?

Reply:
I dont believe I contributed this, I think it appears green because i moved this stuff around. I don't know if its actually needed.